File: Sources/GatewayCore/PartialDownloadRecovery.swift
import Foundation

/// What a resumable installer can prove about its persistent partial file
/// before it makes a network request.
public enum PartialDownloadState: Equatable, Sendable {
    case missing
    case resumable(Int64)
    case complete
    case oversized(Int64)
}

public enum PartialDownloadRecovery {
    public static func inspect(_ url: URL, expectedBytes: Int64,
                               fileManager: FileManager = .default) -> PartialDownloadState {
        guard fileManager.fileExists(atPath: url.path) else { return .missing }
        let attributes = try? fileManager.attributesOfItem(atPath: url.path)
        let bytes = (attributes?[.size] as? NSNumber)?.int64Value ?? 0
        if bytes == expectedBytes { return .complete }
        if bytes > expectedBytes { return .oversized(bytes) }
        return .resumable(bytes)
    }
}

File: cross-platform/src/core/partialDownloadRecovery.ts
/**
 * What a resumable installer can prove about its persistent partial file
 * before it makes a network request, ported from `PartialDownloadRecovery.swift`.
 */
import { existsSync, statSync } from "fs";

export type PartialDownloadState =
  | { kind: "missing" }
  | { kind: "resumable"; bytes: number }
  | { kind: "complete" }
  | { kind: "oversized"; bytes: number };

export function inspect(path: string, expectedBytes: number): PartialDownloadState {
  if (!existsSync(path)) return { kind: "missing" };
  // Existence and size are two separate questions on the Swift side too: a
  // file that exists but whose attributes cannot be read (permissions, a
  // race) reads as zero bytes, not as missing.
  let bytes = 0;
  try { bytes = statSync(path).size; } catch { /* stays 0, matching Swift's `?? 0` */ }
  if (bytes === expectedBytes) return { kind: "complete" };
  if (bytes > expectedBytes) return { kind: "oversized", bytes };
  return { kind: "resumable", bytes };
}
